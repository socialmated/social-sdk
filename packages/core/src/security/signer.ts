import { type Options } from 'got';

export interface Signer<T> {
  sign: ((req: Options) => Promise<T>) & ((req: Request) => Promise<T>);
}
