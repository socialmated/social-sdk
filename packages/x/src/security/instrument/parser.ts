import { gotScraping } from '@social-sdk/client/http';
import { type AnonymousFunctionDeclaration, type FunctionDeclaration, Parser, type ReturnStatement } from 'acorn';
import { simple } from 'acorn-walk';

/**
 * Represents a parsed function containing the source code and its AST node.
 */
export interface ParsedFunction {
  source: string;
  node: FunctionDeclaration;
}

/**
 * Parses the UI metrics JavaScript file to extract the target function that computes UI metrics.
 */
export class UIMetricsJsParser {
  /**
   * The target function that contains the UI metrics logic.
   */
  private targetFunction: FunctionDeclaration | null = null;

  /**
   * Constructs a new instance of the parser and parses the provided script text.
   *
   * @param sourceCode - The JavaScript source code to be parsed.
   */
  constructor(private sourceCode: string) {}

  /**
   * Creates a UIMetricsJsParser by fetching and parsing the UI metrics JavaScript.
   *
   * @returns A promise that resolves to an instance of UIMetricsJsParser.
   */
  public static async create(): Promise<UIMetricsJsParser> {
    const jsFileContent = await gotScraping('https://twitter.com/i/js_inst?c_name=ui_metrics', {
      headers: {
        accept: '*/*',
        referer: 'https://x.com/',
        'sec-fetch-dest': 'script',
        'sec-fetch-mode': 'no-cors',
        'sec-fetch-site': 'cross-site',
      },
    }).text();

    return new UIMetricsJsParser(jsFileContent);
  }

  /**
   * Parses the source code to locate and extract the target function's source as a string.
   *
   * @returns The source code of the target function if found, or null if not found.
   */
  public parseTargetFunction(): ParsedFunction | null {
    const ast = Parser.parse(this.sourceCode, {
      ecmaVersion: 2020,
      sourceType: 'script',
      allowReturnOutsideFunction: true,
    });

    // Walk the AST to find the target function
    simple(ast, {
      FunctionDeclaration: (node) => {
        this.checkFunction(node);
      },
    });

    if (!this.targetFunction) {
      return null;
    }

    return {
      source: this.sourceCode.slice(this.targetFunction.start, this.targetFunction.end),
      node: this.targetFunction,
    };
  }

  /**
   * Checks whether the given function node contains a direct return statement
   * that returns an object with 'rf' and 's' properties.
   *
   * @param node - The function declaration node to check.
   */
  private checkFunction(node: FunctionDeclaration | AnonymousFunctionDeclaration): void {
    // Look for function that has a return statement with an object containing 'rf' and 's' properties
    let hasTargetReturn = false;
    for (const stmt of node.body.body) {
      if (stmt.type === 'ReturnStatement' && this.isTargetReturnStatement(stmt)) {
        hasTargetReturn = true;
        break;
      }
    }

    if (hasTargetReturn) {
      this.targetFunction = node as FunctionDeclaration;
    }
  }

  /**
   * Determines whether a given return statement node returns an object expression
   * containing both 'rf' and 's' properties.
   *
   * @param returnNode - The AST node representing a return statement.
   * @returns `true` if the return statement returns an object with both 'rf' and 's' properties; otherwise, `false`.
   */
  private isTargetReturnStatement(returnNode: ReturnStatement): boolean {
    const returnArg = returnNode.argument;

    // Check if return statement returns an object expression
    if (!returnArg || returnArg.type !== 'ObjectExpression') {
      return false;
    }

    const properties = returnArg.properties;

    // Look for 'rf' and 's' properties
    const hasRfProperty = properties.some(
      (prop) => prop.type === 'Property' && prop.key.type === 'Literal' && prop.key.value === 'rf',
    );
    const hasSProperty = properties.some(
      (prop) => prop.type === 'Property' && prop.key.type === 'Literal' && prop.key.value === 's',
    );

    return hasRfProperty && hasSProperty;
  }
}
