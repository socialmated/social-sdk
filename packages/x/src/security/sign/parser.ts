import { JSDOM, type BinaryData } from 'jsdom';
import { HeaderGenerator } from 'header-generator';
import { gotScraping } from 'got-scraping';

/**
 * Parses X (formerly Twitter) homepage HTML to extract key information required for signing transactions.
 *
 * @remarks
 * - The parser relies on specific HTML patterns present in the X homepage.
 * - If the homepage structure changes, the parser may need to be updated accordingly.
 *
 * @public
 */
class HomeHtmlParser {
  /**
   * Regular expression to extract the ondemand JavaScript file chunk hash from the homepage HTML.
   * The hash is used to construct the URL for fetching the ondemand file.
   */
  private static readonly ON_DEMAND_FILE_REGEX = /['"]ondemand\.s['"]:\s*['"](?<hash>\w*)['"]/;

  /**
   * The document object representing the homepage HTML.
   * This is used to query and manipulate the DOM for extracting information.
   */
  private homeDocument: Document;

  /**
   * Constructs a new parser instance from the homepage HTML.
   * @param htmlText - The HTML content of X's homepage as a string, Buffer, or BinaryData.
   */
  constructor(htmlText: string | Buffer | BinaryData) {
    this.homeDocument = new JSDOM(htmlText).window.document;
  }

  /**
   * Creates a new instance of {@link HomeHtmlParser} by fetching the homepage HTML from x.com.
   *
   * @returns A promise that resolves to a {@link HomeHtmlParser} instance containing the fetched homepage HTML.
   * @throws If the homepage cannot be fetched successfully.
   */
  public static async create(): Promise<HomeHtmlParser> {
    const homeHtml = await gotScraping('https://x.com/home').buffer();
    return new HomeHtmlParser(homeHtml);
  }

  /**
   * Retrieves SVG animation frame elements from the homepage HTML.
   * Frames are identified by element IDs starting with 'loading-x-anim'.
   *
   * @returns Array of SVG frame elements.
   */
  public getAnimationFrames(): SVGElement[] {
    return Array.from(this.homeDocument.querySelectorAll("[id^='loading-x-anim']"));
  }

  /**
   * Locates and fetches the ondemand JavaScript file referenced in the homepage HTML.
   * The file is required for extracting key byte indices used in transaction ID generation.
   *
   * @returns The content of the ondemand file as a string.
   * @throws Error if the ondemand file cannot be found or fetched.
   */
  public findOnDemandJsChunkHash(): string {
    // Serialize the homepage HTML for regex searching.
    const responseStr = this.homeDocument.documentElement.outerHTML;

    // Extract the chunk hash for the ondemand file.
    const onDemandFileMatch = HomeHtmlParser.ON_DEMAND_FILE_REGEX.exec(responseStr);
    const chunkHash = onDemandFileMatch?.groups?.['hash'];
    if (!chunkHash) {
      throw new Error("Couldn't extract ondemand file chunk hash from the page source");
    }

    return chunkHash;
  }
  /**
   * Extracts the verification key from the homepage HTML.
   * The verification key is stored in a meta tag with the name 'twitter-site-verification'.
   *
   * @returns The verification key as a string.
   * @throws Error if the verification key cannot be found.
   */
  public getVerificationKey(): string | null {
    // Find the meta tag containing the verification key.
    const element = this.homeDocument.querySelector("[name='twitter-site-verification']");
    if (!element) {
      throw new Error("Couldn't find meta tag with name 'twitter-site-verification'");
    }

    return element.getAttribute('content');
  }
}

/**
 * Parses a remote ondemand JavaScript file to extract cryptographic key byte indices.
 *
 * @remarks
 * - The ondemand JavaScript file is fetched from a Twitter CDN using a provided chunk hash.
 * - The indices are extracted using a regular expression and are expected to change frequently.
 *
 * @public
 */
class OnDemandJsParser {
  /**
   * Regular expression to extract key byte indices from the ondemand JavaScript file.
   * The indices are used for cryptographic key derivation.
   */
  private static readonly INDICES_REGEX = /\(\w\[(?<index>\d{1,2})\],\s*16\)/g;

  /**
   * Constructs a new parser instance from the ondemand JavaScript file content.
   * @param sourceCode - The JavaScript source code of the ondemand file as a string.
   */
  constructor(private sourceCode: string) {}

  /**
   * Creates an instance of `OnDemandJsParser` by fetching and parsing a remote JavaScript file
   * identified by the provided `chunkHash`.
   *
   * @param chunkHash - The hash string used to construct the URL for the ondemand JavaScript file.
   * @returns A promise that resolves to an `OnDemandJsParser` instance containing the fetched JavaScript content.
   * @throws If the fetch request fails or the response is not OK.
   */
  public static async create(chunkHash: string): Promise<OnDemandJsParser> {
    const jsResponse = await fetch(`https://abs.twimg.com/responsive-web/client-web/ondemand.s.${chunkHash}a.js`, {
      // Requests will be blocked if user-agent is unknown.
      headers: new HeaderGenerator().getHeaders(),
    });
    if (!jsResponse.ok) {
      throw new Error(`Failed to fetch ondemand file: ${jsResponse.statusText}`);
    }

    const jsContent = await jsResponse.text();

    return new OnDemandJsParser(jsContent);
  }

  /**
   * Extracts key byte indices from the ondemand file content.
   * These indices are used for cryptographic key derivation.
   * Keep in mind that these indices changes frequently.
   *
   * @param onDemandFile - The text content of the ondemand file. If omitted, uses the last fetched file.
   * @returns Tuple of [rowIndex, keyByteIndices], where keyByteIndices is an array of three numbers.
   * @throws Error if the key byte indices cannot be extracted.
   */
  public getIndices(): [number, number, number, number] {
    const keyByteIndices: number[] = [];

    // Extract all index matches using the regex.
    const matches = Array.from(this.sourceCode.matchAll(OnDemandJsParser.INDICES_REGEX));
    for (const match of matches) {
      const index = match.groups?.['index'];
      if (!index) {
        throw new Error("Couldn't extract index from regex match");
      }
      keyByteIndices.push(parseInt(index, 10));
    }

    return keyByteIndices as [number, number, number, number];
  }
}

export { HomeHtmlParser, OnDemandJsParser };
