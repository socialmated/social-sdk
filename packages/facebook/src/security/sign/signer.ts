import { type Signer } from '@social-sdk/client/security';
import { type Options } from '@social-sdk/client/http';
import { generateProof, type AppSecretProof } from './sign.js';

export class AppSecretProofSigner implements Signer<AppSecretProof> {
  constructor(private readonly appSecret: string) {}

  public sign(req: Options): AppSecretProof {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- should be defined in the request
    const accessToken = new URLSearchParams(new URL(req.url!).searchParams).get('access_token');
    if (!accessToken) {
      throw new Error('Access token is required to generate App Secret Proof');
    }

    return generateProof(this.appSecret, accessToken);
  }
}
