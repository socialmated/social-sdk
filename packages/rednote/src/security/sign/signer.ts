import { type CookieSession } from '@social-sdk/core/auth/session';
import { type Options } from 'got';
import { type Signer } from '@social-sdk/core/security';
import { generateLocalId } from '../token/token.js';
import { signOld, signNew, generateMns, type XhsSignOutput } from './sign.js';
import { defaultConfig } from '@/client/config.js';

/**
 * The output type for the XHS signing process.
 */
interface XhsSignerOptions {
  /**
   * The type of signing to perform.
   */
  type?: 'sign_new' | 'sign_old';

  /**
   * Whether MNS is supported.
   */
  mns?: boolean;
}

/**
 * The `XhsSigner` class provides methods to sign requests for the Xiaohongshu (XHS) API.
 *
 * @example
 * ```typescript
 * const signer = XhsSigner.fromSession(session);
 * const signature = await signer.sign('/api/resource', requestBody, { type: 'sign_new' });
 * ```
 */
class XhsSigner implements Signer<{ 'X-s': string; 'X-t': string; 'X-Mns': string }> {
  /**
   * The `a1` cookie value.
   */
  private a1: string;

  /**
   * Creates an instance of `XhsSigner`.
   * @param a1 - The `a1` cookie value used for signing requests.
   */
  constructor(a1?: string) {
    this.a1 = a1 ?? generateLocalId(defaultConfig.platform);
  }

  /**
   * Creates an instance of `XhsSigner` from a given session.
   *
   * @param session - The `CookieSession` object containing authentication cookies.
   * @returns An instance of `XhsSigner` initialized with the `a1` cookie value from the session.
   * @throws If the session does not contain an `a1` cookie.
   */
  public static fromSession(session: CookieSession): XhsSigner {
    const a1 = session.get('a1');
    if (!a1) {
      throw new Error('Session does not contain a1 cookie');
    }

    return new XhsSigner(a1);
  }

  /**
   * Generates a signature for the given request.
   *
   * @param request - The request options.
   * @param options - Optional parameters for signing, including type and MNS support.
   * @returns An object containing the signed headers, including 'X-s', 'X-t', and 'X-Mns'.
   */
  public sign(request: Options, options?: XhsSignerOptions): XhsSignOutput {
    const fullPath = request.url ? new URL(request.url).pathname + new URL(request.url).search : '';
    const body: unknown =
      !request.json && typeof request.body === 'string' ? JSON.parse(request.body) : (request.json ?? request.body);
    let output: Omit<XhsSignOutput, 'X-Mns'>;

    if (options?.type === 'sign_old') {
      output = signOld(fullPath, body);
    } else {
      output = signNew(fullPath, body, this.a1);
    }

    return {
      ...output,
      'X-Mns': options?.mns ? generateMns(fullPath, body) : 'unload',
    };
  }
}

export { XhsSigner };
export type { XhsSignerOptions };
