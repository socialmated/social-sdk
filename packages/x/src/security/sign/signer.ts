import { type Options } from 'got';
import { type Signer } from '@social-sdk/core/security';
import { HomeHtmlParser, OnDemandJsParser } from './parser.js';
import { transactionId, deriveAnimationKey, decodeVerificationKey } from './sign.js';

/**
 * Generates and signs the transaction ID for X's API requests.
 *
 * @example
 * ```typescript
 * const signer = new TransactionIdSigner();
 * const transactionId = await signer.sign('POST', '/i/api/1.1/some-endpoint.json');
 * console.log(transactionId);
 * ```
 *
 * @public
 */
class TransactionIdSigner implements Signer<string> {
  /**
   * The cryptographic verification key extracted from the homepage.
   */
  private key?: Uint8Array;

  /**
   * The animation key derived from homepage animation data.
   */
  private animationKey?: string;

  /**
   * Generates and signs a transaction ID for a given HTTP method and API endpoint path.
   *
   * @param method - The HTTP method (e.g., 'GET', 'POST').
   * @param path - The API endpoint path (e.g., '/api/endpoint').
   * @param options - The got Options object.
   * @returns Promise resolving to a Base64-encoded transaction ID string.
   * @throws Error if the signer has not been initialized.
   */
  public async sign(method: string, path: string): Promise<string>;
  /**
   * Generates and signs a transaction ID for a `got` request {@link Options}.
   *
   * @param opts - The `got` request {@link Options} object.
   * @returns Promise resolving to a Base64-encoded transaction ID string.
   * @throws Error if the signer has not been initialized.
   */
  public async sign(opts: Options): Promise<string>;
  /**
   * Generates and signs a transaction ID for a `fetch` {@link Request}.
   *
   * @param req - The `fetch` request {@link Request} object.
   * @returns Promise resolving to a Base64-encoded transaction ID string.
   * @throws Error if the signer has not been initialized or if invalid arguments are provided.
   */
  // eslint-disable-next-line @typescript-eslint/unified-signatures -- explicitly overload
  public async sign(req: Request): Promise<string>;
  public async sign(arg1: string | Options | Request, arg2?: string): Promise<string> {
    if (!this.key || !this.animationKey) {
      await this.init();
    }

    let method: string;
    let requestPath: string;
    if (typeof arg1 === 'string') {
      method = arg1;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- expected to be defined at this point
      requestPath = arg2!;
    } else {
      method = arg1.method;
      requestPath = arg1.url ? new URL(arg1.url).pathname : '';
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- expected to be defined at this point
    return transactionId(method, requestPath, this.key!, this.animationKey!);
  }

  /**
   * Initializes the instance by extracting cryptographic keys and animation data.
   * This method must be called before signing transaction IDs.
   *
   * @returns Promise that resolves when initialization is complete.
   * @throws Error if required data cannot be extracted.
   */
  private async init(): Promise<void> {
    if (this.key !== undefined && this.animationKey !== undefined) {
      // Already initialized
      return;
    }

    // Create parsers for the homepage and ondemand JavaScript file.
    const htmlParser = await HomeHtmlParser.create();
    const jsParser = await OnDemandJsParser.create(htmlParser.findOnDemandJsChunkHash());

    // Extract the verification key from the homepage.
    const verificationKey = htmlParser.getVerificationKey();
    if (!verificationKey) {
      throw new Error("Couldn't extract verification key from the homepage");
    }
    this.key = decodeVerificationKey(verificationKey);

    // Derive the animation key from SVG frames and specific indices.
    const frames = htmlParser.getAnimationFrames();
    const indices = jsParser.getIndices();
    this.animationKey = deriveAnimationKey(this.key, indices, frames);
  }
}

export { TransactionIdSigner };
