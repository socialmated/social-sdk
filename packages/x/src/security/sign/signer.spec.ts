import { readFileSync } from 'node:fs';
import path from 'node:path';
import { TransactionIdSigner } from './signer.js';
import { HomeHtmlParser, OnDemandJsParser } from './parser.js';

describe('TransactionIdSigner', () => {
  const homeHtml = readFileSync(path.join(import.meta.dirname, '__fixtures__', 'home.html'), 'utf-8');
  const onDemandFile = readFileSync(path.join(import.meta.dirname, '__fixtures__', 'ondemand.s.57b4929a.js'), 'utf-8');

  describe('sign', () => {
    it('should generate a transaction ID', async () => {
      vi.spyOn(Math, 'random').mockReturnValue(0.5);
      vi.setSystemTime(Date.UTC(2025, 0, 1));

      vi.spyOn(HomeHtmlParser, 'create').mockResolvedValue(new HomeHtmlParser(homeHtml));
      vi.spyOn(OnDemandJsParser, 'create').mockResolvedValue(new OnDemandJsParser(onDemandFile));

      const signer = new TransactionIdSigner();
      const transactionId = await signer.sign('POST', '/i/api/1.1/some-endpoint.json');

      expect(transactionId).toMatchSnapshot();
    });
  });
});
