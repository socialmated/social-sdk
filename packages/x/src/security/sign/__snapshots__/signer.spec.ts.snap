// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionIdSigner > sign > should sign request and return transaction ID 1`] = `"gLJqtv8MTyPOysRGn4mtXOL7q5dOGkOK4K3QKPY5/P5QvUHSu0Wk2tU+W1w+Y0JgdpCipYPdBuFw3fUCyRJpPO5HkBlvgw"`;
